bin/asset-pipeline

```diff
@@ -0,0 +1,400 @@
+    const outputFile = path.join(this.dest, hashedPath);
+
+    try {
+      await fs.ensureFile(outputFile);
+      await fs.writeFile(outputFile, minified.code);
+      console.log(
+        `Writing minified JS to ${this.getPathRelativeToCWD(outputFile)}`
+      );
+      this.updateKeys({ key, fileHash, written: true });
+    } catch (error) {
+      console.error(chalk.red(`Error writing generated JS: ${error}`));
+    }
+  }
+
+  async rewriteHTML(origPath, key, assetMap = this.assetMap) {
+    let text = await fs.readFile(origPath, 'utf8');
+    let replacementsCount = 0;
+
+    const postcssOptions = { from: origPath };
+    const postcssfilterType = /^text\/css$/;
+
+    const output = await posthtml()
+      .use(
+        posthtmlCSS(
+          [
+            postcssURL({
+              url: (asset) => {
+                if (this.hasHashedReplacementURL(asset.url, assetMap)) {
+                  replacementsCount++;
+                  return this.getReplacementURL(asset.url, assetMap);
+                } else {
+                  return asset.url;
+                }
+              },
+            }),
+          ],
+          postcssOptions,
+          postcssfilterType
+        )
+      )
+      .use(
+        posthtmlURL({
+          eachURL: (url, attr, element) => {
+            if (this.hasHashedReplacementURL(url, assetMap)) {
+              replacementsCount++;
+              return this.getReplacementURL(url, assetMap);
+            } else {
+              return url;
+            }
+          },
+        })
+      )
+      .process(text);
+
+    const fileHash = this.getContentHash(output.html);
+
+    if (replacementsCount) {
+      console.log(
+        `Re-writing ${replacementsCount} ${
+          replacementsCount == 1 ? 'path' : 'paths'
+        } in ${this.getPathRelativeToCWD(origPath)}`
+      );
+    }
+
+    let outputFile;
+    const ext = path.extname(origPath);
+    if (ext !== '.html') {
+      outputFile = path.join(this.dest, this.getHashedPath(key, fileHash));
+    } else {
+      outputFile = path.join(this.dest, path.relative(this.src, origPath));
+    }
+
+    try {
+      await fs.ensureFile(outputFile);
+      await fs.writeFile(outputFile, output.html);
+      const fileType = ext.replace(/^\./, '');
+
+      console.log(
+        `Writing ${fileType} to ${this.getPathRelativeToCWD(outputFile)}`
+      );
+      this.updateKeys({ key, fileHash, written: true });
+    } catch (error) {
+      console.error(chalk.red(`Error writing generated ${fileType}: ${error}`));
+    }
+  }
+
+  async rewriteCSS(origPath, key, assetMap = this.assetMap) {
+    let css = await fs.readFile(origPath, 'utf8');
+    let replacementsCount = 0;
+    const output = await postcss()
+      .use(
+        postcssURL({
+          url: (asset) => {
+            let assetURL = asset.url;
+
+            // This is a special case to support relative paths in CSS files.
+            if (assetURL.startsWith('../')) {
```

Comment:
💣

Reply:
I added test coverage, I think it's pretty okay for now. That will depend on the new mockups to implement too.